rqt_gui_py/src/rqt_gui_py/rclpy_spinner.py

```diff
@@ -23,17 +27,56 @@ def __init__(self, node):
         super().__init__()
         self._node = node
         self._abort = False
+        self._listener_queue = Queue()
 
     def run(self):
         qDebug('Start called on RclpySpinner, spinning ros2 node')
         executor = MultiThreadedExecutor()
         executor.add_node(self._node)
         while rclpy.ok() and not self._abort:
             executor.spin_once(timeout_sec=1.0)
+
+            # Remove subscribers and clients that have been registered for destruction
+            while not self._listener_queue.empty():
+                listener = self._listener_queue.get()
+                self.__destroy_listener__(listener)
+
         if not self._abort:
             qWarning('rclpy.shutdown() was called before QThread.quit()')
 
     def quit(self):  # noqa: A003
         qDebug('Quit called on RclpySpinner')
         self._abort = True
         super().quit()
+
+    def register_listeners_for_destruction(self, *listeners):
```

Comment:
This needs very prominent documentation. It is very reasonable for users to remove subscribers themselves if they don't know about this.

Reply:
Any suggestions? I imagine they will find rqt_gui crashing look at how other plugins do this

Reply:
Correction, spin is thread safe but destroying subscribers is not.